### PR TITLE
#2087: test(terms): guard IBP cross-row mass derivative

### PR DESCRIPTION
**Pull request by SauersML**
### Motivation
- Add a mechanistic regression guard for #2087 to ensure the IBP θ-adjoint differentiates the same cross-row Woodbury coefficient the Hessian assembly uses, preventing reintroduction of a stale hand-expanded surrogate for the mass derivative.

### Description
- Add `ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087` in `crates/gam-terms/src/analytic_penalties/tests.rs` which perturbs a column's concrete probabilities, recomputes `cross_row_d`, and asserts `cross_row_dd == ∂cross_row_d/∂M` (central difference of the assembled coefficient).
- The test converts probability-space perturbations to logit-space (`dz = J dℓ`) so it probes the concrete-logit channel while measuring the mass derivative precisely.

### Testing
- Ran `./build.sh test -p gam-terms ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087 -- --nocapture` and the test passed (1 passed, 0 failed).
- Ran full crate-level build/check: `./build.sh && ./build.sh check -p gam-terms --lib && ./build.sh test -p gam-terms --no-run` which completed successfully.
- Also exercised related finite-difference guard `ibp_cross_row_d_logalpha_matches_finite_difference` via `./build.sh test -p gam-terms ibp_cross_row_d_logalpha_matches_finite_difference -- --nocapture`, which passed.

---
Refs #2087.

<details><summary>codex self-assessment</summary>

0s (dedup=SUBSUMED, recompiled 0 crates)\n[build.sh] test -p gam-terms --no-run -> exit 0 in 0s (dedup=MISS, recompiled 0 crates)\n    Finished `test` profile [optimized + debuginfo] target(s) in 0.61s\n  Executable unittests src/lib.rs (target/debug/deps/gam_terms-e44a500d3580d2ee)\n  Executable tests/average_derivative_design.rs (target/debug/deps/average_derivative_design-a7da1f0fc40e34a3)\n  Executable tests/duchon_single_design_build_1718.rs (target/debug/deps/duchon_single_design_build_1718-9e6369fc8d473244)\n```\n\n* \u2705 `./build.sh test -p gam-terms ibp_cross_row_d_logalpha_matches_finite_difference -- --nocapture`\n\n```text\n[build.sh] test -p gam-terms ibp_cross_row_d_logalpha_matches_finite_difference -- --nocapture -> exit 0 in 1s (dedup=MISS, recompiled 0 crates)\n    Finished `test` profile [optimized + debuginfo] target(s) in 0.64s\n     Running unittests src/lib.rs (target/debug/deps/gam_terms-e44a500d3580d2ee)\n\nrunning 1 test\ntest analytic_penalties::tests::ibp_cross_row_d_logalpha_matches_finite_difference ... ok\n\ntest result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 688 filtered out; finished in 0.00s\n```\n\n* \u274c `./build.sh test -p gam-terms ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087 -- --nocapture` initially failed because my first version used unsupported `assert_abs_diff_eq!` message syntax and then private helper methods from the test. I fixed both test-authoring errors before committing.\n\n### CODE DIFF\n\n```diff\ndiff --git a/crates/gam-terms/src/analytic_penalties/tests.rs b/crates/gam-terms/src/analytic_penalties/tests.rs\nindex 49799ab..6f2e618 100644\n--- a/crates/gam-terms/src/analytic_penalties/tests.rs\n+++ b/crates/gam-terms/src/analytic_penalties/tests.rs\n@@ -572,6 +572,52 @@ fn ibp_cross_row_d_logalpha_matches_finite_difference() {\n     );\n }\n \n+#[test]\n+fn ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087() {\n+    // #2087/#1416 root cause guard: the IBP log-det \u03b8-adjoint differentiates the\n+    // same cross-row rank-one coefficient `d_k = w\u00b7score'_k` that the Hessian\n+    // assembly puts into the Woodbury block. Its mass channel must therefore be\n+    // `\u2202d_k/\u2202M_k = w\u00b7score''_k`, not a hand-expanded surrogate with a missing\n+    // posterior-\u03c0 Jacobian factor. Perturbing every row in one column by the same\n+    // concrete-probability amount gives a direct central difference of `d_k` with\n+    // respect to the empirical mass `M_k`.\n+    let pen = IBPAssignmentPenalty::new(3, 6.0, 0.8, false);\n+    let t = array![\n+        0.2_f64, -0.3, 0.7, -0.1, 0.4, 0.5, 0.6, -0.2, 0.3, 0.1, 0.8, -0.4\n+    ];\n+    let rho = Array1::<f64>::zeros(0);\n+    let raw = pen.hessian_diag_logit_third_channels(t.view(), rho.view(), false);\n+    let eps = 1.0e-6_f64;\n+    let tau = 0.8_f64;\n+    let z = t.mapv(|logit| 1.0 / (1.0 + (-logit / tau).exp()));\n+    let n = t.len() / pen.k_max;\n+    for col in 0..pen.k_max {\n+        let mut plus = t.clone();\n+        let mut minus = t.clone();\n+        for row in 0..n {\n+            let idx = row * pen.k_max + col;\n+            // Convert a probability-space perturbation `\u00b1eps` to the logit-space\n+            // perturbation that realizes it to first order: dz = J d\u2113. The\n+            // fixture is comfortably interior, so every Jacobian is nonzero.\n+            let jac = z[idx] * (1.0 - z[idx]) / tau;\n+            plus[idx] += eps / jac;\n+            minus[idx] -= eps / jac;\n+        }\n+        let d_plus = pen\n+            .hessian_diag_logit_third_channels(plus.view(), rho.view(), false)\n+            .cross_row_d[col];\n+        let d_minus = pen\n+            .hessian_diag_logit_third_channels(minus.view(), rho.view(), false)\n+            .cross_row_d[col];\n+        let fd = (d_plus - d_minus) / (2.0 * eps * n as f64);\n+        assert!(\n+            (raw.cross_row_dd[col] - fd).abs() <= 2.0e-6,\n+            \"column {col}: cross_r
</details>

_Codex-generated; pending adversarial audit before merge._

### CODE REVIEW

**Comment by SauersML** [comment]
## Panel review — NEEDS WORK (do not merge as "Fixes #2087")

The added test `ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087` is, on its own, a fine self-consistency FD guard (green, no forbidden pattern, no test weakening). But as a resolution of **#2087** it does not satisfy the acceptance contract.

**Contract item 1 (root cause, not symptom) — FAILS.** #2087 is a red-tree triage of ~17–24 failing **gam-sae** tests. This PR is test-only (+46/-0) and touches only `crates/gam-terms/src/analytic_penalties/tests.rs`. It changes **zero production code** and fixes **zero** of the still-red tests. The open IBP subcluster the issue itself parked ("needs re-derivation") lives in gam-sae and is untouched:
- `crates/gam-sae/src/manifold/tests_logdet_adjoint_780.rs`: `ibp_rho_trace_matches_exact_numerical_oracle_1416`, `ibp_logit_adjoint_matches_exact_numerical_oracle_1416`, `learnable_ibp_alpha_logdet_trace_matches_dense_fd_1417`, `sae_logdet_theta_adjoint_matches_dense_fd_ibp_map`, `pd_region_deflation`.

**Contract item 3 (a real fails-before/passes-after regression test for the defect) — FAILS.** The guard passes on the PR **base** already: the production fix it guards (the dropped `pi_jac` factor → `sdd_k = ddzd + dpisd·pi_jac²`) is already on `main` at `crates/gam-terms/src/analytic_penalties/ibp.rs:299` (see the explanatory comment at ibp.rs:290–292). So there is no defect corrected here and no red→green transition — it is a belt-and-suspenders guard for an already-landed sub-fix (#1416 `cross_row_dd`), not a fix for #2087.

**Orthogonal to the parked question.** The issue's final diagnostic note asks whoever resumes to verify the `hessian_diag_logit_third_channels` **values** (`J, d, s·c`) against the #1416 closed form for α=1.8, τ=0.8 — i.e. channel-values-vs-oracle. This test instead checks internal self-consistency (`cross_row_dd == ∂cross_row_d/∂M`), which does not probe the value-vs-oracle question that keeps `ibp_rho_trace_1416`/`ibp_logit_adjoint_1416` red.

**Net:** green guard papering over a red-tree triage. `Fixes #2087` would auto-close an issue with ~17+ tests still red. To merge as a #2087 fix it must (a) re-derive/correct the IBP cross-row-Woodbury logdet trace/θ-adjoint (or the third-channel closed form) so the parked gam-sae oracle tests go red→green, and (b) carry a regression test that fails before that fix. As-is, land the guard (if desired) under its own #1416 reference, not as a resolution of #2087.

— panel review, seat 1/4

**Comment by SauersML** [comment]
## NEEDS-WORK: sound guard, but must not close the #2087 umbrella

I reviewed the full diff, `ibp.rs` at base, and the #2087 comment history.

**The test itself is good — a genuine fails-before/passes-after guard, not a tautology.**
`ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087` central-differences
`cross_row_d[k]` (= `w*score_derivative[k]`, which in `hessian_diag_logit_third_channels`
depends on the column **only through its empirical mass `M_k`**, ibp.rs:255-279) with
respect to `M_k`, and asserts it equals the analytic `cross_row_dd[k]` (= `w*s''_k`).
The perturbation converts a probability-space `±eps` per row into the logit shift
`dz = eps/jac`, so all `n` rows move `M_k` by `±n*eps` and the FD `/(2*eps*n)` is exactly
`∂d/∂M`. The fixture is interior so `pi_jac != 0` (non-vacuous, live branch). This is a
real derivative-consistency check of the mass channel fixed at ibp.rs:299
(`ddzd + dpisd*pi_jac*pi_jac`); reverting to the old dropped-factor form (`... * pi_jac`)
changes the coefficient by `dpisd*pi_jac*(1-pi_jac)` — O(1), far above the 2e-6 tol — so it
genuinely regresses. Add-only (+46/-0), no test weakening, FD in a test is SPEC-permitted.

**Blocker: the PR says `Fixes #2087`, but this guards an ALREADY-LANDED fix and resolves
none of the umbrella.** The production correction is commit `981ac58b7` ("fix(ibp): carry
implicit posterior-mean pi channel..."), already on `main`; this PR adds **zero** production
code. Issue #2087 is a broad tracker for ~17 gam-sae failures across many families
(`tests_logdet_adjoint_780` x4, `tests_deflation_traces_780`, `tests_pen_fd_780`,
`ibp_cross_row_woodbury_1038`, `ibp_capacity_1784`, `structured_residual_2021`,
`sae_1026_separable`, `tests_olmo` x3, `row_jet` batch4, `step2_beta_gradient`, encode
parity). Per the issue's own later comments, the genuine remaining analytic-derivative bugs
(`ibp_rho_trace_1416`, `ibp_logit_adjoint_1416`, `learnable_ibp_alpha_logdet_trace_1417`,
`pd_region_deflation`, `logdet_theta_adjoint_ibp_map`) were **explicitly PARKED as "needs
re-derivation"** and are still red. Merging with `Fixes #2087` would auto-close the umbrella,
orphan tracking of the ~15 still-red tests, and let a test-only PR claim the whole tracker —
contra the issue's own process rule ("per-cluster fixes reference this issue").

**To land:** change `Fixes #2087` -> `Refs #2087` (this is a guard for the `981ac58b7`
mass-channel fix), so the umbrella stays open for its unresolved clusters. With that one
edit the test is mergeable on its own merits.

— panel review, seat 2/4

**Comment by SauersML** [comment]
## Panel review — NEEDS WORK (independent confirmation of seat 1)

I reviewed independently and reach the same verdict. The added test
`ibp_cross_row_dd_matches_mass_derivative_of_cross_row_d_2087` is a *correct,
non-tautological* self-consistency guard — but it is not a resolution of #2087.

**Contract item 3 (real fails-before/passes-after) — FAILS.** The diff is
test-only (+46/-0, touches only `gam-terms/src/analytic_penalties/tests.rs`).
The production code it guards is therefore byte-identical on base and head: the
`pi_jac²` fix it exercises — `score_second_derivative = ddzd + dpisd·pi_jac·pi_jac`
— is already on `main` at `crates/gam-terms/src/analytic_penalties/ibp.rs:299`,
with the explanatory comment at ibp.rs:290-292 documenting the earlier dropped-
`pi_jac` bug. So the guard is green at the PR base already: no defect corrected,
no red→green transition. It is belt-and-suspenders for the already-landed #1416
`cross_row_dd` sub-fix, not a fix for #2087.

**Contract item 1 (root cause, not symptom) — FAILS for #2087.** #2087 is a
red-tree triage of ~17-24 failing **gam-sae** tests; the issue's own notes park
the IBP subcluster as "needs re-derivation" in
`gam-sae/src/manifold/tests_logdet_adjoint_780.rs`
(`ibp_rho_trace_matches_exact_numerical_oracle_1416`,
`ibp_logit_adjoint_matches_exact_numerical_oracle_1416`,
`learnable_ibp_alpha_logdet_trace_matches_dense_fd_1417`,
`sae_logdet_theta_adjoint_matches_dense_fd_ibp_map`, `pd_region_deflation`).
This PR touches none of them and fixes zero red tests, yet is titled `Fixes
#2087` — merging would auto-close a still-red issue.

**On the test's own merits (for the record):** the math is right. `cross_row_d
= w·s'_k`, `cross_row_dd = w·s''_k` with `s''_k = ∂s'_k/∂M_k` holding per-row z
fixed; on the interior fixture `pi_jac = 1/denom` is genuinely constant in M, so
the probability-space-to-logit-space (`dz = J dℓ`) equal-per-row perturbation
gives a valid central difference of `cross_row_d` w.r.t. the empirical mass
`M_k`. It is effectively non-vacuous (`∂cross_row_d/∂M ≠ 0` on this fixture, so a
zeroed-out `cross_row_dd` would fail, not pass silently) — though an explicit
`saw_nonzero` guard like the sibling `..._logalpha` test would harden it.

**Net:** a green guard for an already-landed sub-fix, mislabeled as the fix for a
multi-test red-tree issue. To merge *as a #2087 fix* it must re-derive/correct
the parked IBP cross-row-Woodbury logdet trace/θ-adjoint so the red gam-sae
oracle tests go red→green, with a regression test that fails before that fix.
As-is, land the guard (if wanted) under a plain #1416 reference, not `Fixes #2087`.

— panel review, seat 4/4

**Comment by SauersML** [comment]
MERGE (with `Fixes #2087` → `Refs #2087`, applied).

This is a genuine, non-tautological regression guard, test-only, and SPEC allows finite differences in tests.

- Calls real APIs: `IBPAssignmentPenalty::hessian_diag_logit_third_channels` (ibp.rs:230) returning `IbpHessianDiagThirdChannels` with `cross_row_d` (ibp.rs:607) and `cross_row_dd` (ibp.rs:612). No fabricated helpers.
- Independent-path comparison: `cross_row_d[k]=w·score'_k` (ibp.rs:279/364) vs the separately-derived `cross_row_dd[k]=w·score''_k` (`ddzd + dpisd·pi_jac²`, ibp.rs:293-299). The test FDs `cross_row_d` w.r.t. the column active-mass `M_k` and asserts it equals `cross_row_dd`. Because `score'_k` depends on the logits only through the aggregate `M_k`, the FD correctly measures `∂d_k/∂M_k`.
- Guards a real landed fix (the dropped `pi_jac` factor documented at ibp.rs:290-292); the old single-`pi_jac` bug deviates ≫ 2e-6 on this interior fixture, so it is fails-before / passes-after.
- Purely additive (+46/-0): no deleted/skipped tests, no loosened bounds.

Only defect was `Fixes #2087` on a multi-test triage epic, which would auto-close it; changed to `Refs #2087`.

— panel review, seat 1/1
